R/derive_param_tte.R

```diff
@@ -0,0 +1,616 @@
+#'   mutate(STUDYID = "AB42")
+#'
+#' ttae <- tte_source(
+#'   dataset_name = "ae",
+#'   date = AESTDTC,
+#'   set_values_to = vars(
+#'     EVENTDESC = "AE",
+#'     SRCDOM = "AE",
+#'     SRCVAR = "AESTDTC",
+#'     SRCSEQ = AESEQ))
+#'
+#' eos <- tte_source(
+#'   dataset_name = "adsl",
+#'   date = EOSDT,
+#'   censor = 1,
+#'   set_values_to = vars(
+#'     EVENTDESC = "END OF STUDY",
+#'     SRCDOM = "ADSL",
+#'     SRCVAR = "EOSDT"))
+#'
+#' derive_param_tte(
+#'   dataset_adsl = adsl,
+#'   by_vars = vars(AEDECOD),
+#'   start_date = TRTSDT,
+#'   event_conditions = list(ttae),
+#'   censor_conditions = list(eos),
+#'   source_datasets = list(adsl = adsl, ae = ae),
+#'   set_values_to = vars(
+#'     PARAMCD = paste0("TTAE", as.numeric(as.factor(AEDECOD))),
+#'     PARAM = paste("Time to First", AEDECOD, "Adverse Event"),
+#'     PARCAT1 = "TTAE",
+#'     PARCAT2 = AEDECOD
+#'   )
+#' ) %>%
+#' select(USUBJID, STARTDT, PARAMCD, PARAM, ADT, CNSR, SRCSEQ)
+derive_param_tte <- function(dataset = NULL,
+                             dataset_adsl,
+                             by_vars = NULL,
+                             start_date = TRTSDT,
+                             start_date_imputation_flag = NULL,
+                             start_time_imputation_flag = NULL,
+                             event_conditions,
+                             censor_conditions,
+                             source_datasets,
+                             create_datetime = FALSE,
+                             set_values_to,
+                             subject_keys = vars(STUDYID, USUBJID)) {
+  # checking and quoting #
+  assert_data_frame(dataset, optional = TRUE)
+  assert_vars(by_vars, optional = TRUE)
+  start_date <- assert_symbol(enquo(start_date))
+  start_date_imputation_flag <- assert_symbol(enquo(start_date_imputation_flag),
+                                              optional = TRUE)
+  start_time_imputation_flag <- assert_symbol(enquo(start_time_imputation_flag),
+                                              optional = TRUE)
+  assert_data_frame(
+    dataset_adsl,
+    required_vars = quo_c(
+      start_date,
+      start_date_imputation_flag,
+      start_time_imputation_flag
+    )
+  )
+  assert_vars(subject_keys)
+  assert_list_of(event_conditions, "tte_source")
+  # check that censor == 0 is used for events (strongly recommended by CDISC) #
+  assert_list_element(list = event_conditions,
+                      element = "censor",
+                      condition = censor == 0,
+                      message_text = "The censor value of events must be zero.")
+  assert_list_of(censor_conditions, "tte_source")
+  # check that censor > 0 is used for censorings (strongly recommended by CDISC) #
+  assert_list_element(list = censor_conditions,
+                      element = "censor",
+                      condition = censor > 0,
+                      message_text = "The censor value of censorings must be positive.")
```

Comment:
Instead of havin one constructor, i.e. `tte_source()`, and checking for valid censor values here, how about we have two constructors, e.g. `censor_source()` and `event_source()`, that perform these checks. Alternatively, we could add an `event_type` parameter to `tte_source()` which is either `"event"` or `"censor"`. This would be more in line with the existing `_source()` constructors which perform all checks in their function body.

Reply:
I have added `censor_source()` and `event_source()`.